### PR TITLE
 Added tissue binary masks to brainweb phantoms generation

### PR DESCRIPTION
**Pull request by snawarhussain**
### Summary

This pull request introduces some enhancements and new features related to binary mask generation for different tissue when downloading data from the BrainWeb website. The key changes are as follows:

1. **Binary Mask Generation**:
    - Binary masks are now created from tissue properties during the data download process from the BrainWeb website using the `generate_brainweb_phantoms` function. These masks accurately represent different tissue types in the downloaded brain phantoms.

2. **Integration with VoxelGrid Phantom and SimData**:
    - The functionality to resample, plot, and manage these binary masks has been integrated into the `VoxelGrid` phantom and `SimData` classes. This integration ensures that the binary masks are now a cohesive part of the simulation data workflow.

3. **New Example Notebook**:
    - A new example notebook has been added at `documentation/playground/templates/binary_masks.ipynb`. This notebook demonstrates how to extract and plot the binary masks, providing users with a clear and practical guide on utilizing this new functionality.

4. **Intended Use of Binary Masks**:
- Image metric calculation i,e (tissue-specific SNR, Sharpness, contrast, etc.)

### Changes Made
- Updated the `generate_brainweb_phantoms` function to create binary masks based on tissue properties during the BrainWeb data download.
- Enhanced the `VoxelGrid` and `SimData` classes to include functionality for resampling, plotting, and managing these binary masks.
- Added a new Jupyter notebook to the `documentation/playground/templates/` directory that shows how to work with these binary masks, including plotting and extraction.
- Updated the relevant documentation to reflect the new and modified features.

### Testing
- I have tested the code on Python 3.10.14 with binary masks of varying interpolated sizes, subjects, and configuration i.e. [3T, 3T-highres-fat ].
- The example notebook includes visualizations that verify the correct creation, resampling, and plotting of the binary masks.

### Notes
- Feedback is welcome, and I am prepared to make any further refinements or adjustments as needed based on the review.

I look forward to your review and comments on this contribution!

### CODE REVIEW

File: python/MRzeroCore/phantom/voxel_grid_phantom.py [hunk 1]
```diff
@@ -406,39 +438,45 @@ def plot(self) -> None:
             print(f"Plotting slice {s} / {self.PD.shape[2]}")
 
         plt.figure(figsize=(12, 10))
-        plt.subplot(331)
+        plt.subplot(431)
```
Comment (J-Endres) [inline python/MRzeroCore/phantom/voxel_grid_phantom.py:441]: For many usecases masks are not needed and the plot should not be changed. It also will not work if there are more than 3 masks to plot.

Instead, create and show a second figure _if_ there are any masks (where the subplot count adjusts to the mask count)

File: python/MRzeroCore/phantom/brainweb/__init__.py [hunk 2]
```diff
@@ -141,6 +141,11 @@ def noise() -> np.ndarray:
                 for alias in brainweb_data["download-aliases"][tissue]
             ])
             tissue_map = downsample(tissue_map, config_data["downsample"])
+            tiusse_mask = tissue_map > 0.2
```
Comment (J-Endres) [inline python/MRzeroCore/phantom/brainweb/__init__.py:144]: Here the masks are binary, but the input segmentation data is float because voxels can be mixtures of multiple tissues at the borders of the tissues.

I don't know the exact application, but it might be more suitable to keep the masks in the [0, 1] range thoughout the code and as float32

Comment (J-Endres) [inline python/MRzeroCore/phantom/brainweb/__init__.py:144]: please rename ``tiusse`` to ``tissue``

File: python/MRzeroCore/phantom/sim_data.py [hunk 3]
```diff
@@ -66,7 +66,8 @@ def __init__(
         dephasing_func: Callable[[torch.Tensor, torch.Tensor], torch.Tensor],
         recover_func: Callable[[SimData], Any] | None = None,
         phantom_motion=None,
-        voxel_motion=None
+        voxel_motion=None,
+        Masks: Optional[Dict[str,torch.Tensor]] = None,
```
Comment (J-Endres) [inline python/MRzeroCore/phantom/sim_data.py:70]: Please change to lower-case. All variables except physical properties (T1, T2 etc)  are ``snake_case``

File: python/MRzeroCore/phantom/sim_data.py [hunk 4]
```diff
@@ -126,7 +128,8 @@ def cuda(self) -> SimData:
             self.dephasing_func,
             self.recover_func,
             self.phantom_motion,
-            self.voxel_motion
+            self.voxel_motion,
+            Masks=self.Masks  # no need to put on cuda since it's not use for simulation rather for image metric calculation
```
Comment (J-Endres) [inline python/MRzeroCore/phantom/sim_data.py:132]: Put it on cuda nevertheless; users might want to calculate the loss on GPU as well (image reconstruction is often done on GPU, not only the simulation)

File: python/MRzeroCore/phantom/sim_data.py [hunk 5]
```diff
@@ -150,7 +153,8 @@ def cpu(self) -> SimData:
             self.dephasing_func,
             self.recover_func,
             self.phantom_motion,
-            self.voxel_motion
+            self.voxel_motion,
+            Masks=self.Masks
```
Comment (J-Endres) [inline python/MRzeroCore/phantom/sim_data.py:157]: See cuda comment - move back to cpu here accordingly

File: python/MRzeroCore/phantom/voxel_grid_phantom.py [hunk 6]
```diff
@@ -97,6 +97,7 @@ def __init__(
         B1: torch.Tensor,
         coil_sens: torch.Tensor,
         size: torch.Tensor,
+        Masks: Optional[Dict[str,torch.Tensor]] = None,
```
Comment (J-Endres) [inline python/MRzeroCore/phantom/voxel_grid_phantom.py:100]: lower case, same as in SimData

File: python/MRzeroCore/phantom/voxel_grid_phantom.py [hunk 7]
```diff
@@ -193,14 +197,23 @@ def load(cls, file_name: str) -> VoxelGridPhantom:
...
-        
+
+            try:
+                Masks = {}
+                tissues = [key[5:] for key in data.keys() if key.startswith('mask_')]
+                for tissue in tissues:
+                    mask = torch.tensor(data[f'mask_{tissue}'], dtype=torch.uint8)
+                    Masks[tissue] = mask
+            except KeyError:
+                print("No masks found")
```
Comment (J-Endres) [inline python/MRzeroCore/phantom/voxel_grid_phantom.py:208]: There can't be a key error when iterating over keys, this code could be shortened to:

```python
masks = {key: torch.tensor(mask) for key, mask in data.items() if key.startswith('mask_')}
```

Also, depending on your choice of data type in the brainweb generation, this might be a float32 (or do not set the data type explicietley)

File: python/MRzeroCore/phantom/voxel_grid_phantom.py [hunk 8]
```diff
@@ -389,6 +420,7 @@ def resample_multicoil(tensor: torch.Tensor) -> torch.Tensor:
             resample_multicoil(self.B1),
             resample_multicoil(self.coil_sens),
             self.size.clone(),
+            Masks= resample_masks(self.Masks)
         )
 
     def plot(self) -> None:
```
Comment (J-Endres) [inline python/MRzeroCore/phantom/voxel_grid_phantom.py:426]: I would prefer an argument `plot_masks=False` that ensures that plotting by default behaves as before

File: documentation/playground/templates/binary_masks.ipynb
Comment (J-Endres) [inline documentation/playground/templates/binary_masks.ipynb:1]: Please remove outputs on Jupyter notebook.
They are generated when building the documentation but not in the git repository to avoid having constantly changing metadata in the history.

The directory of this file might not be final as the documentation is in need of an overhaul, but might be ok for now.

**Review by J-Endres — CHANGES_REQUESTED** [review]

**Comment by J-Endres** [comment]
Thank you for your contribution!

Your changes look like a gread addition to MRzero and I'm happy to see it getting more practical for more usecases. There are only some small tweaks I'd like to see so that existing behaviour does not change, otherwise I'm more than happy to merge your modifications.

**Comment by snawarhussain** [comment]
Hi @J-Endres , thank you for going through the PR in such great depth, I appreciate the insights and suggestions.
I will make the changes over the weekend and let you know :)

File: python/MRzeroCore/phantom/voxel_grid_phantom.py (hunk 1 above)
Comment (snawarhussain) [reply python/MRzeroCore/phantom/voxel_grid_phantom.py]: I have made the suggested changes and now extra rows(s) are added according to the number of masks and whether the `plot_masks` flag is `True` or `False`

File: python/MRzeroCore/phantom/brainweb/__init__.py (hunk 2 above)
Comment (snawarhussain) [reply python/MRzeroCore/phantom/brainweb/__init__.py]: I understand, and realize that masks can be converted to binary later on at the user end with their preferred thresholding and use. In this case, there is no reason to add a new `tissue_mask` item to dict and we can use the `tissue_map` ( the tissue segmentation map ranging [0,1] which is already part of the data, so I just added it into the `voxel_grid_phantom`  subsequent build process. I have already made these changes. Please have a look and let me know if there are any other changes that need to be made.

Comment (snawarhussain) [reply python/MRzeroCore/phantom/brainweb/__init__.py]: this also has been taken care of

File: python/MRzeroCore/phantom/sim_data.py (hunk 3 above)
Comment (snawarhussain) [reply python/MRzeroCore/phantom/sim_data.py]: the lower case `masks` was a bit confusing since `voxel_grid_phantom` also uses `mask` for thresholding on `PD` so i changed the `Masks` into `tissue_masks`

File: python/MRzeroCore/phantom/sim_data.py (hunk 4 above)
Comment (snawarhussain) [reply python/MRzeroCore/phantom/sim_data.py]: Done,

File: python/MRzeroCore/phantom/sim_data.py (hunk 5 above)
Comment (snawarhussain) [reply python/MRzeroCore/phantom/sim_data.py]: Done

File: python/MRzeroCore/phantom/voxel_grid_phantom.py (hunk 6 above)
Comment (snawarhussain) [reply python/MRzeroCore/phantom/voxel_grid_phantom.py]: done

File: python/MRzeroCore/phantom/voxel_grid_phantom.py (hunk 7 above)
Comment (snawarhussain) [reply python/MRzeroCore/phantom/voxel_grid_phantom.py]: thank you, the latest changes use keys that start with `tissue_`  (tissue_gm, tissue_wm, etc) without explicit datacasting. the output masks are tensors of float32

File: python/MRzeroCore/phantom/voxel_grid_phantom.py (hunk 8 above)
Comment (snawarhussain) [reply python/MRzeroCore/phantom/voxel_grid_phantom.py]: Done

**Comment by snawarhussain** [comment]
The last few forces pushed were amending the commits, making minor adjustments to the comments and the titles on the ipynb example file, and accounting for the number of masks. though the brainweb phantom comes with 3 segmentation masks but i understand that the `VoxelGridPhantom` class intends to be a bit universal and should work on various types of digital phantoms and segmentation masks.

File: python/MRzeroCore/phantom/sim_data.py (hunk 3 above)
Comment (J-Endres) [reply python/MRzeroCore/phantom/sim_data.py]: that's even better

**Review by J-Endres — APPROVED** [review]

**Comment by J-Endres** [comment]
The changes look great, thank you!
